### DISCUSSION TITLE
Cannot upgrade Xamarin to MAUI - VS 2022 for Mac

### DISCUSSION THREAD

**Discussion by inimirpaz** [dotnet/maui#16701] [Upgrading from Xamarin]
### Description

Hi everyone, I guess I am in the wrong github but I do not know where else to ask for help.

Has anyone else upgraded anything else than a Xamarin sample project (I am not even able to do that actually) to .NET MAUI successfully on MacOS?

Any help would be appreciated. I also tried upgrading the project on Windows and bring it back to MacOS: https://github.com/dotnet/upgrade-assistant/issues/1502

My currently installed workloads are:
```
maui-maccatalyst                    7.0.92/7.0.100              SDK 7.0.300               
maui-ios                            7.0.92/7.0.100              SDK 7.0.300               
maui-android                        7.0.92/7.0.100              SDK 7.0.300               
ios                                 16.4.7089/7.0.100           SDK 7.0.300               
maui                                7.0.92/7.0.100              SDK 7.0.300               
android                             33.0.68/7.0.100             SDK 7.0.300      
```

### Steps to Reproduce

Not able to use the upgrade assistant on MacOS. Upgrading on Windows does not generate a working project (besides the manual operations needed) and bringing it back to MacOS the project does not open correctly on VS for Mac because of missing workloads (even thought they are installed)

### Link to public reproduction project repository

none

### Version with bug

Unknown/Other

### Last version that worked well

Unknown/Other

### Affected platforms

I was *not* able test on other platforms

### Affected platform versions

none

**Comment by JoacimWall** [comment]
Hi
If it’s a large solution I would create a new project an start move pice by pice. 
by doing this you will get single project solution, global using and easy to find and fix errors. If you convert all at ones you will have hard time to figure out what issue is generate errors. 
One other thing is that you just copy you files to the now solution with finder not using the vs for Mac add files. If you view the project file you will se the it’s adding rows that are not needed. 

There would be a Mac version of the upgrade assisted that you run from command line if I remember correct.  You can use this to get all the namespaces correct before you start moving code. I used visual studio code to fix namespace changes with find and replace. 

I would move small part of code and build so you don’t spend all days to try figure out build errors.  look in the build log not the error window where it’s shows errors that is not the root issue.

**Comment by inimirpaz** [comment]
Hi @JoacimWall ,

First of all thank you for your reply! Very much appreciated.

> There would be a Mac version of the upgrade assisted that you run from command line if I remember correct. You can use this to get all the namespaces correct before you start moving code. I used visual studio code to fix namespace changes with find and replace.

Yes, the upgrade assistant is the one I referenced in my opening post, but it does not work because it asks for the workloads to be installed even though they are.

> I would move small part of code and build so you don’t spend all days to try figure out build errors. look in the build log not the error window where it’s shows errors that is not the root issue.

Thanks, I have just started the procedure and I am doing as you say.

I am a bit worried and dissatisfied with the upgrading procedure though, which at this point is more of a migration than an upgrade if you will... due to the breaking changes with quite a lot of "old" NuGet packages and such.
Speaking of which, how do you handle non-upgraded NuGets to .NET MAUI? Is doing the manual work of downloading the source code of a NuGet and retarget it for .net6+ the only viable option?

Furthermore, one last question: since what .NET MAUI asks is basically a migration, is there a full list of what to change to be MAUI-style instead of Xamarin-style? You mentioned the single project solution (which is a great thing) and global usings.
I am looking for all the things I should do MAUI-style (e.g. I know there is something new related to IoC, Dependency Injection, Messaging, MVVM, Commands, Converters, Renderers ... to name a few). All I can find are bits and pieces, I would love a compendium of everything to keep in mind when moving to .NET MAUI and HOW to do it properly (not just a short bullet summary) to avoid to build an hybrid xamarin&maui as a code base for the next years to come.

Thank you all!

I keep this ticket open just in case anyone else wants to contribute, and/or help to fix the upgrade-assistant thing.

**Comment by JoacimWall** [comment]
Hi 
I don't know of any compendium but I would suggest that you look in to the nuget CommunityToolkit.Mvvm. I use it the for Dependency Injection and Messaging. 
if you need to resolve different service that not are injected in the constructor this helper could be nice. Think it was David Ortinau who showed this at some point.

 ```
public static class ServiceHelper
{
    public static TService GetService<TService>() => Current.GetService<TService>();

    public static IServiceProvider Current =>
#if ANDROID
        MauiApplication.Current.Services;
#elif IOS
        MauiUIApplicationDelegate.Current.Services;
#else
        null;
#endif

}

Using it
var srv =  ServiceHelper.Current.GetService<INotificationService>();
```

Regarding Nugets Most of the nugets that I use has bin updated to DotNet 6 and works for me. Those that was not I have bin able to replace for others. It's way too much work to fix nugets and migrate at the same time. If there is something specific nuget that you can't live without, you can email maui-upgrades@microsoft.com and maybe they will help you.

Another important question regarding this is that Msoft prioritizes fixing most of the bugs in Net 8 which will come in November, so if you need to have a version in production before then and can't use preview, I would wait as I think there are too many problems with Net 7 and that it is not prioritized by Msoft. We have a smaller app that we will release with version 7. But also a bigger app that we will wait to release until 8 is ready. So right now we have to code two versions of the bigger app one which is Xamarin and one MAUI this probably creates an additional cost of around 25%.

//Joacim

**Comment by davidortinau** [comment]
In addition to the other comments...

We shipped a new .NET Upgrade Assistant this week, including a fixed macOS CLI version if you want to go back and try that. In any case, almost never with the output of the assistant leave you in a state to compile successfully. That's the manual steps. 

The recommended approach above to create a new project and copy over your code is great. Many devs have good success doing that. 

I also recommend not adopting new .NET MAUI concepts at first unless you have a really compelling reason. Get your code over and building first. Taking on too much change at once can really slow your progress if not kill it. Copy all your platform specific stuff like app icons, splash screens, images, etc into the Platforms folders and then later you can adopt the MauiImage and other new conveniences. 

Docs have the most information here: https://learn.microsoft.com/en-us/dotnet/maui/migration/

> Speaking of which, how do you handle non-upgraded NuGets to .NET MAUI? Is doing the manual work of downloading the source code of a NuGet and retarget it for .net6+ the only viable option?

Some libraries have been abandoned so I'd recommend replacing them with something that is actively maintained. Or takeover the repos and add the net6-*  and newer TFMs and implementations where necessary.

Reminder, as any gets stuck or have an urgent question, please email us at maui-upgrades@microsoft.com.

**Comment by inimirpaz** [comment]
Hi @davidortinau , thanks for your insight.
I gave the new Upgrade Assistant (version 0.5.261.51392) a try and it is somewhat better, but I still end up in a project which I do not know how to fix to make it build and not because of the typical manual steps which I would expect.

I launched the upgrade assistant three times (Android, iOS and Forms projects respectively) making an in-place upgrade (no errors there). But when I open the solution I get the aforementioned missing workloads errors:

`/usr/local/share/dotnet/sdk/6.0.105/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.ImportWorkloads.targets(5,5): Error NETSDK1147: To build this project, the following workloads must be installed: maui-android
To install these workloads, run the following command: dotnet workload install maui-android (NETSDK1147) (MyApp)`
`/usr/local/share/dotnet/sdk/6.0.105/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.ImportWorkloads.targets(5,5): Error NETSDK1147: To build this project, the following workloads must be installed: maui-ios
To install these workloads, run the following command: dotnet workload install maui-ios (NETSDK1147) (MyApp)`

and VS is asking to install the workloads, which if I do and restart VS, it keeps asking the workloads to be installed again and I cannot get out of the loop.

I must be doing something wrong. I do expect the migration to have manual steps (as explained in the migration guides you linked) but I expect the VS solution to "work" properly in the new MAUI way and to detect the properly installed workloads to the least.

May I ask you, given a non-trivial Xamarin.Forms app project, what you would do to have it migrated to .NET MAUI with the Upgrade Assistant once you -in MacOS from a terminal- have changed directory to the .sln file location and run `upgrade-assistant upgrade`?
In which order would you run the procedure among the projects? Should it be an in-place upgrade or not? I would like to have the project upgraded to the point where I then could finally be tackling the real migration manual steps and errors, not the sln/csproj files not being properly upgraded.

Thank you in advance.

**Comment by inimirpaz** [comment]
@jamesmontemagno to the rescue: https://www.youtube.com/watch?v=7EaHKGUCIqc , this might help. Let's give it a try, at least now we know how the upgrade assistant is supposed to be used.
